frontend/controller/service-worker.js

```diff
@@ -244,22 +165,76 @@ sbp('sbp/selectors/register', {
     const result = await pwa.deferredInstallPrompt.prompt()
     return result.outcome
   }
-})
-
-// helper method
-
-function urlBase64ToUint8Array (base64String) {
-  // reference: https://gist.github.com/Klerith/80abd742d726dd587f4bd5d6a0ab26b6
-  const padding = '='.repeat((4 - (base64String.length % 4)) % 4)
-  const base64 = (base64String + padding)
-    .replace(/-/g, '+')
-    .replace(/_/g, '/')
-
-  const rawData = window.atob(base64)
-  const outputArray = new Uint8Array(rawData.length)
+});
+
+// Events that need to be relayed to the SW
+[LOGIN_COMPLETE, NEW_CHATROOM_UNREAD_POSITION, SET_APP_LOGS_FILTER].forEach((event) =>
+  sbp('okTurtles.events/on', event, (...data) => {
+    navigator.serviceWorker.controller?.postMessage({ type: 'event', subtype: event, data })
+  })
+)
+
+const swRpc = (() => {
+  if (!navigator.serviceWorker) {
+    throw new Error('Missing service worker object')
+  }
+  let controller: ?ServiceWorker = navigator.serviceWorker.controller
+  navigator.serviceWorker.addEventListener('controllerchange', (ev: Event) => {
+    controller = (navigator.serviceWorker: any).controller
+  }, false)
+
+  return (...args) => {
+    return new Promise((resolve, reject) => {
+      if (!controller) {
+        reject(new Error('Service worker not ready'))
+        return
+      }
+      const messageChannel = new MessageChannel()
+      messageChannel.port1.addEventListener('message', (event: MessageEvent) => {
+        if (event.data && Array.isArray(event.data)) {
+          const r = deserializer(event.data[1])
+          // $FlowFixMe[incompatible-use]
+          if (event.data[0] === true) {
+            resolve(r)
+          } else {
+            reject(r)
+          }
+          messageChannel.port1.close()
+        }
+      }, false)
+      messageChannel.port1.addEventListener('messageerror', (event: MessageEvent) => {
+        reject(event.data)
+        messageChannel.port1.close()
+      }, false)
```

Comment:
I'm not sure about this bit of feedback from Claude as I don't know the semantics of event listeners as it relates to MessageChannels:

> The message channel event listeners should be removed when they're no longer needed. While the port is closed, it's good practice to explicitly remove the listeners.

Reply:
I was taking heap snapshots (with Chrome) to compare and test this and I don't think it makes a big difference. If you want to, I could remove the listeners, since it shouldn't introduce _new_ issues. However,

1. Closing the port doesn't seem to be enough for getting the handler garbage collected. In this sense, the suggestion would improve things.
2. However, the port being GC'ed (e.g., by going out of scope) _does_ get handlers and all associated objects cleared.

So, while removing the handler could potentially be helpful, the larger issue would be ensuring that the message port doesn't have any dangling references. AFAICT, these are ephemeral objects that aren't preserved.